contrib/dump-mtd-ifd.py

```diff
@@ -0,0 +1,177 @@
+        flash_descriptor_regs.append(
+            struct.unpack_from(
+                "<I", blob_ifd, offset=flash_region_base_addr + (region * 4)
+            )[0]
+        )
+    for region in range(num_regions):
+        print(f"flash_descriptor_reg{region}=0x{flash_descriptor_regs[region]:X}")
+
+    # parse each partition
+    fregs: List[IfdPartition] = []
+    for i in range(num_regions):
+        freg_base: int = (flash_descriptor_regs[i] << 12) & 0x07FFF000
+        freg_limit: int = ((flash_descriptor_regs[i] >> 4) & 0x07FFF000) | 0x00000FFF
+
+        # invalid
+        if freg_base > freg_limit:
+            continue
+
+        fregs.append(
+            IfdPartition(region=i, offset=freg_base, size=freg_limit - freg_base)
+        )
+
+    # create a binary blob big enough
+    image_size: int = 0
+    for freg in fregs:
+        if freg.offset + freg.size > image_size:
+            image_size = freg.offset + freg.size
+    print(f"image_size=0x{image_size:x}")
+    blob: bytearray = bytearray(image_size)
+
+    # copy each partition
+    for freg in fregs:
+        print("reading...", freg)
+        try:
+            f.seek(freg.offset)
+            blob[freg.offset : freg.offset + freg.size] = f.read(freg.size)
+        except OSError as e:
+            print(f"failed to read: {e}")
+    return blob
+
+
+def _read_device_to_file(devname: str, filename: Optional[str]) -> None:
+    # grab system info from sysfs
+    if not filename:
+        filename = ""
+        for sysfs_fn in [
+            "/sys/class/dmi/id/sys_vendor",
+            "/sys/class/dmi/id/product_family",
+            "/sys/class/dmi/id/product_sku",
+        ]:
+            try:
+                with open(sysfs_fn, "rb") as f:
+                    if filename:
+                        filename += "-"
+                    filename += (
+                        f.read().decode().replace("\n", "").replace(" ", "_").lower()
+                    )
+            except FileNotFoundError:
+                pass
+        if filename:
+            filename += ".bin"
+        else:
+            filename = "bios.bin"
+
+    # check this device name is what we expect
+    print(f"checking /dev/{devname}...")
+    with open(f"/sys/class/mtd/{devname}/name", "rb") as f_name:
+        name = f_name.read().decode().replace("\n", "")
+    if name != "BIOS":
+        print(f"Not Intel Corporation PCH SPI Controller, got {name}")
+        sys.exit(1)
+
+    # read the IFD header, then each partition
+    try:
+        with open(f"/dev/{devname}", "rb") as f_in:
+            print(f"reading from /dev/{devname}...")
+            blob = _read_partitions(f_in)
+    except PermissionError as e:
+        print(f"cannot read mtd device: {e}")
+        sys.exit(1)
+    print(f"writing {filename}...")
+    with open(filename, "wb") as f_out:
+        f_out.write(blob)
+    print("done!")
+
+
+if __name__ == "__main__":
+    # both have defaults
+    parser = argparse.ArgumentParser()
```

Comment:
missing a description of what the tool is supposed to do.

Reply:
Fixed.